include/votca/xtp/openmp_cuda.h

```diff
@@ -58,23 +71,83 @@ class OpenMP_CUDA {
     return 0;
 #endif
   }
+
+  // 3c multiply
   void setOperators(const std::vector<Eigen::MatrixXd>& tensor,
                     const Eigen::MatrixXd& rightoperator);
   void MultiplyRight(Eigen::MatrixXd& matrix);
 
+  // 3c
   void setOperators(const Eigen::MatrixXd& leftoperator,
                     const Eigen::MatrixXd& rightoperator);
   void MultiplyLeftRight(Eigen::MatrixXd& matrix);
 
+  // RPA
   void createTemporaries(Index rows, Index cols);
-  void A_TDA(const Eigen::MatrixXd& matrix, const Eigen::VectorXd& vec);
+  void PushMatrix(Eigen::MatrixXd& mat);
+  void A_TDA(const Eigen::VectorXd& vec);
+
+  // Hd + Hqp + Hd2
+  void createTemporaries(const Eigen::VectorXd& vec,
+                         const Eigen::MatrixXd& input, Index rows1, Index rows2,
+                         Index cols);
+  void PrepareMatrix1(Eigen::MatrixXd& mat);
+  void SetTempZero();
+  void PrepareMatrix2(const Eigen::Block<const Eigen::MatrixXd>& mat, bool Hd2);
+  void Addvec(const Eigen::VectorXd& row);
+  void MultiplyRow(Index row);
+
+  // Hx
+
+  void createAdditionalTemporaries(Index rows, Index cols);
+  void PushMatrix1(Eigen::MatrixXd& mat);
+  void MultiplyBlocks(const Eigen::Block<const Eigen::MatrixXd>& mat, Index i1,
+                      Index i2);
+
   Eigen::MatrixXd getReductionVar();
 
  private:
-  const Eigen::MatrixXd* rightoperator_ = nullptr;
-  const Eigen::MatrixXd* leftoperator_ = nullptr;
+  template <class T>
+  class DefaultReference {
+   public:
+    DefaultReference() = default;
+    DefaultReference(T object) : p(&object){};
+
+    DefaultReference& operator=(const T& object) {
+      p = &object;
+      return *this;
+    }
+
+    const T& operator()() {
+      assert(p != nullptr && "Dangling reference!");
+      return *p;
+    }
+
+   private:
+    const T* p = nullptr;
+  };
```

Comment:
Yes, no. Defaultable references are really evil and they sort of break any intuition about these kind of references. The only reason I use it here, is because the lifetimes are already difficult to understand and it makes the CPU and GPU code a lot more similar.